applications/CoSimulationApplication/custom_python/add_custom_io_to_python.cpp

```diff
@@ -0,0 +1,367 @@
+            numNodesPerElem[elem_counter++] = r_geom.PointsNumber();
+            for (const auto& r_node : r_geom) {
+                elems.push_back(r_node.Id());
+            }
+        }
+    } else {
+        for (const auto& r_elem : rModelPart.Elements()) {
+            const auto& r_geom = r_elem.GetGeometry();
+            numNodesPerElem[elem_counter++] = r_geom.PointsNumber();
+            for (const auto& r_node : r_geom) {
+                elems.push_back(r_node.Id());
+            }
+        }
+    }
+
+    EMPIRE_API_sendMesh(const_cast<char*>(rName.c_str()), numNodes, numElems, &nodes[0], &nodeIDs[0], &numNodesPerElem[0], &elems[0]);
+}
+
+void sendMesh_DefaultName(ModelPart& rModelPart, const bool UseConditions)
+{
+    sendMesh(rModelPart, const_cast<char*>(rModelPart.Name().c_str()), UseConditions);
+}
+
+void recvMesh(ModelPart& rModelPart, const std::string rName, const bool UseConditions)
+{
+    KRATOS_ERROR_IF(rModelPart.NumberOfNodes() > 0) << "ModelPart is not empty, it has nodes!" << std::endl;
+    KRATOS_ERROR_IF(rModelPart.NumberOfProperties() > 0) << "ModelPart is not empty, it has properties!" << std::endl;
+    KRATOS_ERROR_IF(rModelPart.IsDistributed()) << "ModelPart cannot be distributed!" << std::endl;
+
+    int numNodes;
+    int numElems;
+    double** nodes = new double*;
+    int** nodeIDs = new int*;
+    int** numNodesPerElem = new int*;
+    int** elem = new int*;
+
+    EMPIRE_API_recvMesh(const_cast<char*>(rName.c_str()), &numNodes, &numElems, nodes, nodeIDs, numNodesPerElem, elem);
+
+    const std::unordered_map<int, std::string> element_name_map = {
+        // {1 , "Element3D1N"}, // does not yet exist
+        {2 , "Element3D2N"},
+        {3 , "Element3D3N"},
+        {4 , "Element3D4N"}
+    };
+    const std::unordered_map<int, std::string> condition_name_map = {
+        {1 , "PointCondition3D1N"},
+        {2 , "LineCondition3D2N"},
+        {3 , "SurfaceCondition3D3N"},
+        {4 , "SurfaceCondition3D4N"}
+    };
+
+    // fill ModelPart with received entities
+    for (int i=0; i<numNodes; ++i) {
+        rModelPart.CreateNewNode((*nodeIDs)[i], (*nodes)[i*3], (*nodes)[i*3+1], (*nodes)[i*3+2]);
+    }
+
+    auto p_props = rModelPart.CreateNewProperties(0);
+
+    int counter=0;
+    for (int i=0; i<numElems; ++i) {
+        const int num_nodes_elem = (*numNodesPerElem)[i];
+        std::vector<ModelPart::IndexType> elem_node_ids(num_nodes_elem);
+        for (int j=0; j<num_nodes_elem; ++j) {
+            elem_node_ids[j] = (*elem)[counter++];
+        }
+        if (UseConditions) {
+            rModelPart.CreateNewCondition(condition_name_map.at(num_nodes_elem), i+1, elem_node_ids, p_props);
+        } else {
+            rModelPart.CreateNewElement(element_name_map.at(num_nodes_elem), i+1, elem_node_ids, p_props);
+        }
+    }
+
+    // deallocating memory
+    // TODO what abt the memory that was allocated inside "EMPIRE_API_recvMesh"?
+    delete [] nodes;
```

Comment:
these are double pointers ! 
for a 2 nodes you should do something like 
```cpp
for i : {1,2}
  delete [] nodes[i]
```

if you just say delete [] nodes you are just deleting the memory for the vector of pointers which are pointing to the nodal coordinates but not the memory for the coordinates themselves.